### PR TITLE
[refactor] Extract Taichi IR compilation from KernelCodeGen 

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = #562 

`KernelCodeGen` should deal with binary kernel generation only. Taichi IR compilation (lowering) is a different story.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by yuanming-hu** [comment]
This also creates a reusable `compile_to_offloads` function for the Metal/OpenGL backend.

File: taichi/program/program.cpp [hunk 1]
```diff
@@ -96,6 +96,18 @@ Program::Program(Arch desired_arch) {
     engine = std::make_unique<AsyncEngine>();
   }
 
+  if (!arch_is_cpu(config.arch)) {
```
Comment (archibate) [inline taichi/program/program.cpp:99]: what's the diff with arch_is_host?

File: taichi/codegen/codegen_cuda.cpp [hunk 2]
```diff
@@ -1,124 +1,7 @@
 // The CUDA backend
 
 #include "codegen_cuda.h"
-#include "taichi/ir/scratch_pad.h"
 
 TLANG_NAMESPACE_BEGIN
 
-void CodeGenCUDA::lower() {
```
Comment (archibate) [inline taichi/codegen/codegen_cuda.cpp:8]: Great! Now if we want to modify lower, we won't forget one in four. Could you also do this for Metal&GL? Or @k-ye and I are expected to do that in another PR?

File: taichi/program/program.cpp (hunk 1 above)
Comment (yuanming-hu) [reply taichi/program/program.cpp:99]: Do you mean `host_arch`?

File: taichi/codegen/codegen_cuda.cpp (hunk 2 above)
Comment (yuanming-hu) [reply taichi/codegen/codegen_cuda.cpp:8]: Let us do those in another PR. It's better to have this done by someone the most familiar with these backends. OpenGL & Metal are not yet integrated into our CI system after all.

**Comment by yuanming-hu** [comment]
(I'm merging this in to rush for tomorrow's meeting with my advisor.)

File: taichi/codegen/codegen_cuda.cpp (hunk 2 above)
Comment (k-ye) [reply taichi/codegen/codegen_cuda.cpp:8]: BTW, `offload()`'s interface is different from the rest, because it returns `OffloadedResult`

https://github.com/taichi-dev/taichi/blob/62e24dd47a1fa223fb31842e4689e7ecadc417c8/taichi/transforms/offload.cpp#L407

Metal used to need the result to decide the global tmps size, but now we just use the fixed `taichi_global_tmp_size`. So this is no longer necessary for me. @archibate do you still need `OffloadedResult`?

---

I remember we briefly touched the idea of having an IR pass manager, along with all kinds of custom pass implementations. I wonder how should we propagate the value of each pass back (if there is one) to the user. The easiest way might be something below:

```cpp
PassManager mgr;
// FooPass doesn't return anything, so we don't need to keep a reference
mgr.add(std::make_unique<FooPass>());
// BarPass returns something, so we need to keep a reference
BarPass bp;
mgr.add(&bp);
mgr.run();
bp.get_whatever_you_need();
```

However, LLVM's pass registration seems to forbid this type of works -- It seems to me that LLVM automatically creates the pass instances for you in the registration handler, so users don't have access to the instance? Hmm, maybe I shouldn't use IR passes in this way?

Comment (yuanming-hu) [reply taichi/codegen/codegen_cuda.cpp:8]: Right - `offload` returns something different. I think the discussions here are related to #689. Sorry, I haven't got a chance to comment there but I'll do after meeting my advisors today.

For LLVM passes: I'm not familiar with the new `llvm::PassManager` since I've been using `llvm::legacy::PassManager`, which allows you to do something like

https://github.com/taichi-dev/taichi/blob/62e24dd47a1fa223fb31842e4689e7ecadc417c8/taichi/jit/jit_arch_cpu.cpp#L235-L247

If we design it this way, the user can easily do something like
```C++
auto handle = createTargetTransformInfoWrapperPass(target_machine->getTargetIRAnalysis());
module_pass_manager.add(handle);
```

and then `handle` is still available to the user.

Comment (k-ye) [reply taichi/codegen/codegen_cuda.cpp:8]: Ah OK. That aligns with what I want, thx!
